src/main/java/org/elasticsearch/index/store/Store.java

```diff
@@ -665,24 +666,30 @@ public String toString() {
 
         public static final MetadataSnapshot EMPTY = new MetadataSnapshot();
 
-        public MetadataSnapshot(Map<String, StoreFileMetaData> metadata) {
+        private Map<String, String> commitUserData;
+
+        public MetadataSnapshot(Map<String, StoreFileMetaData> metadata, Map<String, String> commitUserData) {
             this.metadata = metadata;
+            this.commitUserData = commitUserData;
         }
 
         MetadataSnapshot() {
-            this.metadata = Collections.emptyMap();
+            this.metadata = Collections.EMPTY_MAP;
+            this.commitUserData = Collections.EMPTY_MAP;
         }
 
         MetadataSnapshot(IndexCommit commit, Directory directory, ESLogger logger) throws IOException {
-            metadata = buildMetadata(commit, directory, logger);
+            final SegmentInfos segmentCommitInfos = Store.readSegmentsInfo(commit, directory);
+            commitUserData = segmentCommitInfos.getUserData();
+            metadata = buildMetadata(commit, directory, logger, segmentCommitInfos);
             assert metadata.isEmpty() || numSegmentFiles() == 1 : "numSegmentFiles: " + numSegmentFiles();
         }
 
-        ImmutableMap<String, StoreFileMetaData> buildMetadata(IndexCommit commit, Directory directory, ESLogger logger) throws IOException {
+        ImmutableMap<String, StoreFileMetaData> buildMetadata(IndexCommit commit, Directory directory, ESLogger logger, SegmentInfos segmentCommitInfos) throws IOException {
             ImmutableMap.Builder<String, StoreFileMetaData> builder = ImmutableMap.builder();
             Map<String, String> checksumMap = readLegacyChecksums(directory).v1();
             try {
-                final SegmentInfos segmentCommitInfos = Store.readSegmentsInfo(commit, directory);
```

Comment:
we can't take this one out of the try catch, it protects agains corruption. I think we should rename `buildMetadata` to `loadMetadata` and set both commutUserData and metadata in it.